File: packages/php/src/Resource/Events.php
<?php

declare(strict_types=1);

namespace LogSeal\Resource;

use LogSeal\LogSeal;

class Events
{
    public function __construct(private readonly LogSeal $client) {}

    /**
     * List events with filtering and pagination.
     *
     * @param array{
     *     organization_id: string,
     *     action?: string,
     *     action_prefix?: string,
     *     actor_id?: string,
     *     target_type?: string,
     *     target_id?: string,
     *     after?: string,
     *     before?: string,
     *     search?: string,
     *     limit?: int,
     *     cursor?: string,
     * } $params
     * @return array Paginated list of events.
     */
    public function list(array $params): array
    {
        return $this->client->request('GET', '/v1/events', null, $params);
    }

    /**
     * Retrieve a single event by ID.
     */
    public function get(string $eventId): array
    {
        return $this->client->request('GET', "/v1/events/{$eventId}");
    }

    /**
     * Verify hash-chain integrity for an organization.
     */
    public function verify(string $organizationId, ?string $after = null, ?string $before = null): array
    {
        $body = ['organization_id' => $organizationId];
        if ($after !== null) $body['after'] = $after;
        if ($before !== null) $body['before'] = $before;
        return $this->client->request('POST', '/v1/events/verify', $body);
    }

    /**
     * Verify a specific sequence range.
     */
    public function verifyRange(string $organizationId, int $fromSequence, int $toSequence): array
    {
        return $this->client->request('POST', '/v1/events/verify-range', [
            'organization_id' => $organizationId,
            'from_sequence' => $fromSequence,
            'to_sequence' => $toSequence,
        ]);
    }

    /**
     * Retrieve the Merkle proof for an event.
     */
    public function getProof(string $eventId): array
    {
        return $this->client->request('GET', "/v1/events/{$eventId}/proof");
    }

    /**
     * Auto-paginate through all matching events.
     *
     * @return \Generator<int, array> Yields individual event arrays.
     */
    public function listAll(array $params): \Generator
    {
        $cursor = null;
        do {
            if ($cursor !== null) {
                $params['cursor'] = $cursor;
            }
            $page = $this->list($params);
            foreach ($page['data'] as $event) {
                yield $event;
            }
            $cursor = $page['next_cursor'] ?? null;
        } while (!empty($page['has_more']) && $cursor !== null);
    }
}

File: packages/ruby/lib/logseal/resources/events.rb
# frozen_string_literal: true

module LogSeal
  module Resources
    class Events
      def initialize(client)
        @client = client
      end

      # List events with filtering and pagination.
      #
      #   page = client.events.list(organization_id: "org_acme", action: "user.login", limit: 50)
      #   page["data"].each { |e| puts e["action"] }
      def list(organization_id:, action: nil, action_prefix: nil, actor_id: nil,
               target_type: nil, target_id: nil, after: nil, before: nil,
               search: nil, limit: nil, cursor: nil)
        params = { organization_id: }.compact
        params[:action] = action if action
        params[:action_prefix] = action_prefix if action_prefix
        params[:actor_id] = actor_id if actor_id
        params[:target_type] = target_type if target_type
        params[:target_id] = target_id if target_id
        params[:after] = after if after
        params[:before] = before if before
        params[:search] = search if search
        params[:limit] = limit if limit
        params[:cursor] = cursor if cursor

        @client.request(:get, "/v1/events", params:)
      end

      # Retrieve a single event by ID.
      def get(event_id)
        @client.request(:get, "/v1/events/#{event_id}")
      end

      # Verify hash-chain integrity.
      def verify(organization_id:, after: nil, before: nil)
        body = { organization_id: }
        body[:after] = after if after
        body[:before] = before if before
        @client.request(:post, "/v1/events/verify", body:)
      end

      # Verify a specific sequence range.
      def verify_range(organization_id:, from_sequence:, to_sequence:)
        @client.request(:post, "/v1/events/verify-range", body: {
          organization_id:, from_sequence:, to_sequence:
        })
      end

      # Retrieve the Merkle proof for an event.
      def get_proof(event_id)
        @client.request(:get, "/v1/events/#{event_id}/proof")
      end

      # Auto-paginate through all matching events.
      #
      #   client.events.list_all(organization_id: "org_acme") do |event|
      #     puts "#{event['action']} by #{event.dig('actor', 'name')}"
      #   end
      #
      # @return [Enumerator] if no block is given.
      def list_all(**params, &block)
        return enum_for(:list_all, **params) unless block

        cursor = nil
        loop do
          page = list(**params, cursor:)
          page["data"].each(&block)
          cursor = page["next_cursor"]
          break unless page["has_more"] && cursor
        end
      end
    end
  end
end
